### PR TITLE
Don't export outside annotations

### LINKED ISSUE #1620: bounding box is in picture when outside property is set
### My actions before raising this issue
- [X] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [X] asked in gitter
- [X] Searched [past issues](/issues)

### Expected Behaviour
The labels should be in the tfrecord as i see them in the browser while labeling. At least what i expect form it. So if i end a tracked label with the "outside property" it also should not be in the exported data or.
or am i doing something wrong? Thanks ;)

### Current Behaviour
(Hi, i don't know if this is a bug or if im doing something wrong but) the bounding box in "track" mode is also in the picture where i set the "outside property" and disappears only in the next frame but I think it should also be not there in the actual frame with "outside property".

### Possible Solution
the export should take the "outside property" into account or the UI should show the bounding box in this frame that we know what we actually label.

### Steps to Reproduce (for bugs)
exported my [data set as CVAT for images 1.1](https://drive.google.com/file/d/1FLwEhJNBWhDioNWtD0cwmtLz8n7pOXP9/view?usp=sharing)
if i export data sets like this i get always one labeled frame more like described.
1. import jpgs extracted from video
2. label bounding boxes in 2 point track mode
3. go to the point wehre de object is no longer in frame and switch to outside property
4. export dataset as TFRecord v1.0
5. look inside the TFRecord file for the frame where the bounding box should be gone with the outside property, but its still there

### Context
i try to correctly label images for export in tfrecord and the train a SSD mobilenet with this data

### Your Environment

- Git hash commit (`git log -1`): fb380d985528409483cfa25550126f9c50a65de4
- Docker version `docker version` (e.g. Docker 17.0.05): 
Bildschirmfoto 2020-05-29 um 20 42 49
Docker on Mac 2.3.0.3 (45519)
- Are you using Docker Swarm or Kubernetes?: Kubernetes is with the Docker but not shure
- Operating System and version (e.g. Linux, Windows, MacOS): MacOS 10.15.4
- Code example or link to GitHub repo or gist to reproduce problem: -
- Other diagnostic information / logs: -

### Next steps
You may [join our Gitter](https://gitter.im/opencv-cvat/public) channel for community support.
already asked there if i did something wrong but no answer thill now

### PR DESCRIPTION
**Pull request by zhiltsov-max**
### Motivation and context

Fixes #1620

`outside` annotations were not considered during export.

### How has this been tested?

Unit tests.

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [x] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by azhavoro** [comment]
@zhiltsov-max could you fix tests?

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

File: cvat/apps/dataset_manager/formats/mot.py [hunk 1]
```diff
@@ -67,6 +67,10 @@ def _import(src_file, task_data):
         for track in tracks.values():
             # MOT annotations do not require frames to be ordered
             track.shapes.sort(key=lambda t: t.frame)
-            # Set outside=True for the last shape in a track to finish the track
-            track.shapes[-1] = track.shapes[-1]._replace(outside=True)
+            # Append a shape with outside=True to finish the track
+            last_shape = track.shapes[-1]
+            if last_shape.frame < int(task_data.meta['task']['stop_frame']):
+                track.shapes.append(last_shape._replace(outside=True,
+                    frame=last_shape.frame + task_data.frame_step)
```
Comment (nmanovic) [inline cvat/apps/dataset_manager/formats/mot.py:74]: I'm not sure that here we need to add task_data.frame_step (probably +1 is correct). You can get a shape with `frame > task_data.meta['task']['stop_frame']`. @azhavoro , could you please look at the line and give your opinion?

File: cvat/apps/dataset_manager/bindings.py [hunk 2]
```diff
@@ -410,9 +410,10 @@ def match_frame(self, filename):
             "Cannot match filename or determine frame number for {} filename".format(filename))
 
 class CvatTaskDataExtractor(datumaro.SourceExtractor):
-    def __init__(self, task_data, include_images=False):
+    def __init__(self, task_data, include_images=False, include_outside=False):
```
Comment (nmanovic) [inline cvat/apps/dataset_manager/bindings.py:413]: Why do we need the flag at all? When is it useful?

File: cvat/apps/dataset_manager/tests/_test_formats.py [hunk 3]
```diff
@@ -401,6 +402,22 @@ def load_dataset(src):
...
+        self._generate_annotations(task)
+        task_ann = TaskAnnotation(task["id"])
+        task_ann.init_from_db()
+        task_data = TaskData(task_ann.ir_data, Task.objects.get(pk=task["id"]))
+
+        extractor = CvatTaskDataExtractor(task_data, include_outside=False)
+        dm_dataset = datumaro.components.project.Dataset.from_extractors(extractor)
+        self.assertEqual(4, len(dm_dataset.get("image_1").annotations))
+
+        extractor = CvatTaskDataExtractor(task_data, include_outside=True)
```
Comment (nmanovic) [inline cvat/apps/dataset_manager/tests/_test_formats.py:417]: I believe we always should include_outside but different formats can use or ignore such shapes.

Comment (zhiltsov-max) [reply cvat/apps/dataset_manager/tests/_test_formats.py:417]: Maybe, but there is no formats with such annotation property.

File: cvat/apps/dataset_manager/formats/mot.py (hunk 1 above)
Comment (zhiltsov-max) [reply cvat/apps/dataset_manager/formats/mot.py:74]: With +1 I'm not sure about how filtered out frames would be handled in UI, e.g. in "go to the next keyframe" button.

File: cvat/apps/dataset_manager/tests/_test_formats.py (hunk 3 above)
Comment (zhiltsov-max) [reply cvat/apps/dataset_manager/tests/_test_formats.py:417]: We've been there already. It means that all formats must expect (read as support) such shapes.

Comment (nmanovic) [reply cvat/apps/dataset_manager/tests/_test_formats.py:417]: Now "caller" knows about format implementation details (a format doesn't support "outside" flag). Is it OK? What is about attributes? Tracks? Cuboids? Are you going to filter that as?

Comment (zhiltsov-max) [reply cvat/apps/dataset_manager/tests/_test_formats.py:417]: I think, only CVAT knows about `outside`. Tracks and cuboids are not exported too.

File: cvat/apps/dataset_manager/formats/mot.py (hunk 1 above)
Comment (nmanovic) [reply cvat/apps/dataset_manager/formats/mot.py:74]: @zhiltsov-max , could you please check?

File: cvat/apps/dataset_manager/tests/_test_formats.py (hunk 3 above)
Comment (nmanovic) [reply cvat/apps/dataset_manager/tests/_test_formats.py:417]: OK. I see that CvatTaskDataExtractor is used as an utility function by each format. Thus the knowledge encapsulated.

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

File: cvat/apps/dataset_manager/formats/mot.py (hunk 1 above)
Comment (zhiltsov-max) [reply cvat/apps/dataset_manager/formats/mot.py:75]: I've checked. We can't just save the next frame (which is `+1`) without breaking the logic in many places, at least on the server, so we probably should stick to `+frame_step`.

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

**Comment by nmanovic** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cvat/apps/dataset_manager/formats/mot.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/cvat/cvat/pullRequest?prid=5708171&bid=18681965)

**Pull request merged**